sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c

```diff
@@ -737,121 +742,186 @@ void gnrc_rpl_recv_DIS(gnrc_rpl_dis_t *dis, kernel_pid_t iface, ipv6_addr_t *src
+ * @param[in] inst      The @p RPL instance that the DIO belongs to.
+ * @param[in] dio       The @p DIO packet.
+ * @param[in] src       The address of the sender.
+ * @param[in] len       The length of the whole DIO packet.
+ * @param[in] is_new    Whether the DIO belongs to an existing or newly created DODAG.
+ *
+ * @returns             True, on success.
+ * @returns             False, if parsing of the options failed.
+ * @returns             False, if the DODAG is new and the GNRC_RPL_OPT_DODAG_CONF option is required but missing.
+ */
+static bool _handle_DIO_opts(gnrc_rpl_instance_t *inst, gnrc_rpl_dio_t *dio, ipv6_addr_t *src,
+                             uint16_t len, bool is_new)
 {
-    (void)dst;
-    gnrc_rpl_instance_t *inst = NULL;
-    gnrc_rpl_dodag_t *dodag = NULL;
+    gnrc_rpl_opt_t *opts = (gnrc_rpl_opt_t *)(dio + 1);
+    uint32_t included_opts = 0;
 
-#ifdef MODULE_NETSTATS_RPL
-    gnrc_rpl_netstats_rx_DIO(&gnrc_rpl_netstats, len, (dst && !ipv6_addr_is_multicast(dst)));
-#endif
+    /* subtract length of ICMPv6 header and DIO base object fields to get length of DIO options */
+    size_t opt_len = len - sizeof(gnrc_rpl_dio_t) - sizeof(icmpv6_hdr_t);
 
-    if (!IS_ACTIVE(CONFIG_GNRC_RPL_WITHOUT_VALIDATION)) {
-        if (!gnrc_rpl_validation_DIO(dio, len)) {
-            return;
-        }
+    if (!_parse_options(GNRC_RPL_ICMPV6_CODE_DIO, inst, opts, opt_len, src, &included_opts)) {
+        DEBUG("RPL: Error encountered during DIO option parsing\n");
+        return false;
     }
 
-    len -= (sizeof(gnrc_rpl_dio_t) + sizeof(icmpv6_hdr_t));
-
-    if (gnrc_rpl_instance_add(dio->instance_id, &inst)) {
-        /* new instance and DODAG */
-        gnrc_netif_t *netif;
-
-        if (byteorder_ntohs(dio->rank) == GNRC_RPL_INFINITE_RANK) {
-            DEBUG("RPL: ignore INFINITE_RANK DIO when we are not yet part of this DODAG\n");
-            gnrc_rpl_instance_remove(inst);
-            return;
-        }
-
-        inst->mop = (dio->g_mop_prf >> GNRC_RPL_MOP_SHIFT) & GNRC_RPL_SHIFTED_MOP_MASK;
-        inst->of = gnrc_rpl_get_of_for_ocp(GNRC_RPL_DEFAULT_OCP);
-
-        if (iface == KERNEL_PID_UNDEF) {
-            netif = _find_interface_with_rpl_mcast();
+    if (is_new && !bit_check32(&included_opts, GNRC_RPL_OPT_DODAG_CONF)) {
+        if (!IS_ACTIVE(CONFIG_GNRC_RPL_DODAG_CONF_OPTIONAL_ON_JOIN)) {
+            DEBUG("RPL: DIO without DODAG_CONF option - request new DIO\n");
+            gnrc_rpl_send_DIS(NULL, src, NULL, 0);
+            return false;
         }
         else {
-            netif = gnrc_netif_get_by_pid(iface);
+            DEBUG("RPL: DIO without DODAG_CONF option - use default trickle parameters\n");
+            gnrc_rpl_send_DIS(NULL, src, NULL, 0);
         }
-        assert(netif != NULL);
+    }
 
-        gnrc_rpl_dodag_init(inst, &dio->dodag_id, netif->pid);
+    return true;
+}
 
-        dodag = &inst->dodag;
+/**
+ * @brief   Updates a DODAG with the info from a recieved DIO packet.
```

Comment:
```suggestion
 * @brief   Updates a DODAG with the info from a received DIO packet.
```

Typo as documented by the static tests.